### ISSUE TITLE
[FR] A freehand drawing board 

### ISSUE THREAD

**Issue by annieappflowy** [AppFlowy-IO/AppFlowy#2091]
### Description

A free-hand sketch/drawing element, just like other elements - text, image, callout, heading etc. Because, no matter how many embeds we offer, sometimes there is something that a user can only write it down by itself and needs to edit it on spot. For this, I even thought of offering an embed to Microsoft Whiteboard, but I prefer it to be something that is a part of the application itself, with features that resonates with other elements. Because, an embed is, still depending on a third-party.

### Impact

Users who express via hand drawing

**Comment by a-wallen** [comment]
Whoever picks this up can refer to [this tutorial](https://www.kodeco.com/25237210-building-a-drawing-app-in-flutter) as an example. 

Just an FYI @annieappflowy, this feature might require some constraints for the following reason: 
Embedding the freehand drawing widget into a document will get a little more tricky because of how Flutter handles [gestures in the gesture arena](https://docs.flutter.dev/development/ui/advanced/gestures#gesture-disambiguation). For example, if a user swipes up inside of the freehand drawing widget, it's hard to disambiguate whether the user's intent is to scroll down on the page or to draw upwards inside of the widget. 

For the aforementioned reason, a reasonable constraint might be that we can embed a AppFlowy FreeHandDrawingWidget in a document, but while it's inside a document, it cannot be editable.

**Comment by rileyhawk1417** [comment]
@a-wallen So to embed the widget per say is like having a separate dedicated part of the document that can be recognised as a drawing area?

**Comment by Xazin** [comment]
The way I see it, whilst embedded in the document, it should be a read-only "area". 

The drawing should happen in an edit/expanded mode. This will also make it easier to build this feature.

What do you think @a-wallen ? I can try to make some quick example in Figma to clarify what I mean if needed.

**Comment by Xazin** [comment]
@a-wallen 
Really bad example, but I didn't want to invest too much time in it, so spent 2 minutes to draw it out in Figma.

![image](https://github.com/AppFlowy-IO/AppFlowy/assets/42929161/23f5ab5e-b53e-42af-8fb3-89bb776a74ae)

Of course when editing versus expanding in read mode, there would be color picker, etc.

**Comment by a-wallen** [comment]
Yep @Xazin, that solution looks good to me: solves the engineering problem with UX :)

**Comment by Xazin** [comment]
Made this, it's not the best, I honestly haven't looked into our designs enough to work according to our design system. But maybe someone can pick it up and make something good off of it.

![image](https://github.com/AppFlowy-IO/AppFlowy/assets/42929161/baf9d566-fb05-457b-b0fd-99f73fcdc27b)

[Figma found here](https://www.figma.com/file/GlyZ9AykmTXop8PiBLHBpo/AppFlowy-Dark-(Copy)?type=design&node-id=2206-382219&t=rIm9y6ca7iHNrGUz-4)

**Comment by Svk1190** [comment]
I think it would be interesting if each document can have two transparent layers, one layer for text (as it is now) and one layer for handwriting / drawing. This would allow users to use their stylus to annotate/highlight/draw on top of their text notes.

The drawing tools can activate the drawing layer. Using the keyboard or the text tools can deactivate the drawing layer and activate the text layer. 

Some specific tools may work on both layers simultaneously, for example, introduction of vertical space, eraser, lasso, etc.

**Comment by imjasoriano** [comment]
for the text layer, I suggest to make the text rotatable and movable so that map roads or diagrams, for example, can easily fit labels. Adding formating to the text would also be nice (ex. font color, size, bold, etc).

**Comment by imjasoriano** [comment]
For the drawing tool, it would be great to have:
1. pens - a few pen options (pen, highlighter, fountain pen), line width option, line color options
2. shapes - resizeable and rotatable shapes like rectangle, oval, triangle, arrow, multi arrow, bent arrow, flexible arrow
3. paper - adjustable page color fill, background image, grid
4. resizable paper size and dimensions

**Comment by hackeresq** [comment]
I would prefer an existing solution - something like Excalidraw. I think making this a block in the editor makes a lot of sense.

**Comment by annieappflowy** [comment]
I would like to make a suggestion, I would love for you to add a whiteboard where you could make diagrams, brainstorm so that we can be more creative with our work. Thank you.

**Comment by SOSeriously** [comment]
I would like to suggest:
1. The whiteboard should be a separate page with its dedicated tools for drawing, creating diagrams, adding images, and adding text. Similar to other whiteboards.
2. To link a whiteboard similar to linking a calendar or kanban board in a document.
3. Drawio has a nice feature called scratchpad where you can save parts of a diagram as a preset, import/export the preset, or add an image from the device as a preset.

**Comment by GOKORURI007** [comment]
This sounds like the edgeless mode in affine.

**Comment by GlassedSilver** [comment]
> This sounds like the edgeless mode in affine.

Yes or OneNote which I'm dying to switch away from.

I NEED handwriting to take notes, because:
- I'm much quicker with handwriting than with typing
- but even so, with handwriting I can much more quickly add a quick doodle to sketch something up to support my thought in-line without placing something which adds a pacing issue when I need to keep up with something

Additionally, even though I so wish to get away from OneNote, it does have some nice tricks up its sleeves like this:
[YouTube demo of synched notes with audio recording](https://youtu.be/IIPbpJWT5ec)

**Comment by gbostock** [comment]
So it's not in the plan? I wish to occasionally doodle while typing.
